[cd] Deploy concourse using instance refresh

philandstuff opened the pull request:
Currently, the deployment pipeline manages to use concourse to deploy
concourse by scaling itself out and in again.  This could be made
simpler by using the new [Instance Refresh feature][1].  This means
that we don't have to change the desired number of workers, or re-run
terraform, or anything.

[1]: https://aws.amazon.com/blogs/compute/introducing-instance-refresh-for-ec2-auto-scaling/

https://trello.com/c/Fkc86srz

philandstuff committed:
[cd] Deploy concourse using instance refresh

Currently, the deployment pipeline manages to use concourse to deploy
concourse by scaling itself out and in again.  This could be made
simpler by using the new [Instance Refresh feature][1].  This means
that we don't have to change the desired number of workers, or re-run
terraform, or anything.

Note: we still have the problem that, while polling the API to see if
all workers have been refreshed, the working doing the polling might
get terminated.  This commit doesn't (yet) fix that, we'll need to
address this.

Note2: there's probably a bunch of files in
reliability-engineering/pipelines/tasks/ that can be deleted now, I
haven't got round to it though.

[1]: https://aws.amazon.com/blogs/compute/introducing-instance-refresh-for-ec2-auto-scaling/

philandstuff commented:
also I'm not sure if the bash conditionals work, I don't know if `[[` supports `-a` like this

philandstuff committed:
wait for instance refresh to finish

Just like with the previous solution, we need to wait until the
instance refresh finishes, and we need to have some way to guard
against the worker we're running on disappearing, so we stick
`attempts: 10` in there.

Krenair commented on reliability-engineering/pipelines/tasks/asg-wait-for-refresh.yml:
```diff
@@ -0,0 +1,42 @@
+      aws autoscaling describe-instance-refreshes \
+        --auto-scaling-group-name "${ASG_NAME}" \
+        --instance-refresh-ids "$(cat refresh/id)" \
+        | jq "InstanceRefreshes[]" \
```
```suggestion
        | jq ".InstanceRefreshes[]" \
```
?

philandstuff commented on reliability-engineering/pipelines/concourse-deployer.yml:
```diff
@@ -169,91 +168,72 @@ jobs:
+    - task: refresh-main-team-workers-asg
+      file: tech-ops/reliability-engineering/pipelines/tasks/asg-refresh-instances.yml
+      output_mapping:
+        refresh: main-team-worker-refresh
```
we write to the `refresh` volume and try to read from it in a separate job.  To fix this, combine the `refresh-instances` and `wait-for-refresh` jobs into one job with different steps.

philandstuff committed:
Fix typo in jq query

Co-authored-by: Alex Monk <alex.monk@digital.cabinet-office.gov.uk>

philandstuff committed:
Merge refresh-instances job with wait-for-refresh

These have to be in the same job because they share
volumes (`web-refresh` and `main-team-worker-refresh`).

philandstuff committed:
refactor refresh-instances

I think it's clearer this way round: we have two in_parallel streams,
one for the web nodes and one for the worker nodes.

philandstuff committed:
Remove unused `get`s of tech-ops-private

We were getting this repo in a bunch of places but we weren't actually
using it.

philandstuff committed:
Remove unused `get`s of concourse-release

We were getting this release in a bunch of places but we weren't
actually using it.

philandstuff commented:
@Krenair updated with feedback, also did some extra tidying commits

risicle commented:
I'd be in favour of this just from the amount of bash it removes.
